### PR TITLE
Feat/keyp 270 native se client service

### PR DESCRIPTION
**Pull request by odelcroi**

### CODE REVIEW

File: java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientServiceFactory.java [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+    private static final Logger logger =
+            LoggerFactory.getLogger(NativeSeClientServiceFactory.class);
+
+    /**
+     * Init the builder
+     * 
+     * @return next configuration step
+     */
+    NodeStep builder() {
```
Comment (andrei-cristea) [inline java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientServiceFactory.java:40]: Cette méthode doit être public

File: java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        return uniqueInstance;
+    }
+
+
+    @Override
+    public <T extends KeypleUserData> T executeRemoteService(RemoteServiceParameters parameters,
+            KeypleUserDataFactory<T> userOutputDataFactory) {
+
+        // check params nullity
+        if (parameters == null || userOutputDataFactory == null) {
```
Comment (andrei-cristea) [inline java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientServiceImpl.java:79]: Utiliser Assert.getInstance()....

File: java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+            // execute dto request locally
+            KeypleMessageDto responseDto = executeLocally(nativeReader, receivedDto);
+
+            // get response dto - send dto response to server
+            receivedDto = node.sendRequest(responseDto);
+        }
+
+        // extract userOutputData
+        KeypleUserData userOutputData =
```
Comment (andrei-cristea) [inline java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientServiceImpl.java:171]: Gérer le cas où action == error

File: java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+        JsonObject body = new JsonObject();
+
+        body.addProperty("userInputData", userInputData.toJson());
+        body.add("readerEvent", KeypleJsonParser.getParser().toJsonTree(readerEvent));
+
+        return new KeypleMessageDto()//
+                .setSessionId(generateSessionId())//
+                .setAction(KeypleMessageDto.Action.READER_EVENT.name())//
+                .setSessionId(generateSessionId())//
```
Comment (andrei-cristea) [inline java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientServiceImpl.java:223]: Supprimer car déjà setté

File: java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientUtils.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    /**
+     * Retrieve the NativeSeClient if initiated
+     * 
+     * @return the singleton instance of the service if instanciated, null instead
+     */
+    static public NativeSeClientService getService() {
+        return NativeSeClientServiceImpl.getInstance();
+    }
+
```
Comment (andrei-cristea) [inline java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/NativeSeClientUtils.java:29]: Rajouter un TODO pour la méthode getAsyncNode()

File: java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/TransmitExecutor.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                body.add("seResponse", KeypleJsonParser.getParser().toJsonTree(e.getSeResponse(),
+                        SeResponse.class));
+            }
+
+            // if an exception occurs, send it into a keypleDto to the Master
+            response = new KeypleMessageDto(keypleMessageDto)
+                    .setAction(KeypleMessageDto.Action.ERROR.name())//
+                    .setErrorCode(KeypleMessageDto.ErrorCode.KeypleReaderIOException.getCode())//
+                    .setErrorMessage(e.getMessage())//
+                    .setBody(body.toString());
```
Comment (andrei-cristea) [inline java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/TransmitExecutor.java:92]: Normalement les champs errorCode et body ne doivent pas être tous les deux settés. A discuter

File: java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/TransmitSetExecutor.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                body.add("seResponse", KeypleJsonParser.getParser().toJsonTree(e.getSeResponse(),
+                        SeResponse.class));
+            }
+
+            // if an exception occurs, send it into a keypleDto to the Master
+            response = new KeypleMessageDto(keypleMessageDto)
+                    .setAction(KeypleMessageDto.Action.ERROR.name())//
+                    .setErrorCode(KeypleMessageDto.ErrorCode.KeypleReaderIOException.getCode())//
+                    .setErrorMessage(e.getMessage())//
+                    .setBody(body.toString());
```
Comment (andrei-cristea) [inline java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/TransmitSetExecutor.java:102]: Même remarque

File: java/component/keyple-plugin/remote-se/nativese/src/test/java/org.eclipse.keyple.plugin.remotese.nativese.impl/AbstractNativeSeServiceTest.java [hunk 8]
```diff
@@ -70,10 +70,10 @@ public void findLocalReader_Found() throws Exception {
      * helpers
      */
 
-    public PluginFactory mockPluginFactory() {
+    static public PluginFactory mockPluginFactory() {
```
Comment (andrei-cristea) [inline java/component/keyple-plugin/remote-se/nativese/src/test/java/org.eclipse.keyple.plugin.remotese.nativese.impl/AbstractNativeSeServiceTest.java:73]: public static

File: java/component/keyple-core/src/main/java/org/eclipse/keyple/core/util/json/KeypleJsonParser.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    private static class GsonParser {
+        private static final KeypleJsonParser INSTANCE = new KeypleJsonParser();
+    }
+
+    public static Gson getParser() {
+        return GsonParser.INSTANCE.getGson();
+    }
+
+    private Gson getGson() {
```
[the remark this answers is not in the document]
Comment (odelcroi) [reply java/component/keyple-core/src/main/java/org/eclipse/keyple/core/util/json/KeypleJsonParser.java:42]: mistake of mine, thanks

File: java/component/keyple-plugin/remote-se/core/src/main/java/org/eclipse/keyple/plugin/remotese/core/KeypleUserData.java [hunk 10]
```diff
@@ -24,10 +23,11 @@
...
 
     /**
-     * Converts the object content to a Map of key-values pair strings.
+     * Converts the object content to a json string
      *
      * @return a not null reference.
      * @since 1.0
      */
-    Map<String, String> toMap();
+    String toJson();
```
[the remark this answers is not in the document]
Comment (odelcroi) [reply java/component/keyple-plugin/remote-se/core/src/main/java/org/eclipse/keyple/plugin/remotese/core/KeypleUserData.java:31]: In practice, the user is the one responsible for serializing and deserializing the object to json, for us, it is just a String. Moreover using Json parser is easier than creating a custom object to map parser.

File: java/component/keyple-plugin/remote-se/core/src/main/java/org/eclipse/keyple/plugin/remotese/core/exception/KeypleDoNotPropagateEventException.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+ * This program and the accompanying materials are made available under the terms of the Eclipse
+ * Public License 2.0 which is available at http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ********************************************************************************/
+package org.eclipse.keyple.plugin.remotese.core.exception;
+
+import org.eclipse.keyple.core.seproxy.exception.KeypleException;
+
+public class KeypleDoNotPropagateEventException extends KeypleException {
```
[the remark this answers is not in the document]
Comment (odelcroi) [reply java/component/keyple-plugin/remote-se/core/src/main/java/org/eclipse/keyple/plugin/remotese/core/exception/KeypleDoNotPropagateEventException.java:19]: It is used in EventFilter if the user does not want the event to be propagated

File: java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/TransmitExecutor.java (hunk 6 above)
Comment (odelcroi) [reply java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/TransmitExecutor.java:92]: KeypleReaderIoException contains SeResponses, then we must send them to the server.

File: java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/TransmitSetExecutor.java (hunk 7 above)
Comment (odelcroi) [reply java/component/keyple-plugin/remote-se/nativese/src/main/java/org/eclipse/keyple/plugin/remotese/nativese/impl/TransmitSetExecutor.java:102]: KeypleReaderIoException contains SeResponses, then we must send them to the server.

**Pull request merged**